artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireCoreConverter.java

```diff
@@ -0,0 +1,572 @@
+import org.apache.activemq.util.MarshallingSupport;
+import org.fusesource.hawtbuf.UTF8Buffer;
+
+public class OpenWireCoreConverter {
+   public static final SimpleString JMS_GROUPSEQ = new SimpleString("JMSXGroupSequence");
+   public static final SimpleString JMS_XGROUPSEQ = new SimpleString("JMSXGroupSeq");
+   public static final SimpleString JMS_GROUPID = new SimpleString("JMSXGroupID");
+   private static final SimpleString AMQP_REPLYTOGROUPID = new SimpleString("JMS_AMQP_ReplyToGroupID");
+   public static final SimpleString JMS_CONTENT_TYPE = new SimpleString("JMS_AMQP_ContentType");
+   private static final SimpleString AMQP_CONTETTYPE = new SimpleString("JMS_AMQP_CONTENT_TYPE");
+
+   public static ICoreMessage toCore(final OpenWireMessage openwireMessage, final CoreMessageObjectPools coreMessageObjectPools)
+         throws Exception {
+      CoreMessage coreMessage = new CoreMessage(-1, openwireMessage.getMessageSize(), coreMessageObjectPools);
+      final SimpleString type = openwireMessage.getSimpleStringProperty(OpenWireMessageConverter.JMS_TYPE_PROPERTY);
+      if (type != null) {
+         coreMessage.putStringProperty(OpenWireMessageConverter.JMS_TYPE_PROPERTY, type);
+      }
+      coreMessage.setDurable(openwireMessage.isDurable());
+      coreMessage.setExpiration(openwireMessage.getExpiration());
+      coreMessage.setPriority(openwireMessage.getPriority());
+      coreMessage.setTimestamp(openwireMessage.getTimestamp());
+      final ActiveMQBuffer openWireMessageBuffer = openwireMessage.getReadOnlyBodyBuffer();
+      final boolean isCompressed = openwireMessage.getBooleanProperty(OpenWireMessageConverter.AMQ_MSG_COMPRESSED);
+      final byte coreType = openwireMessage.getType();
+      coreMessage.setType(coreType);
+      openWireMessageBuffer.resetReaderIndex();
+      final byte[] bytes;
+      bytes = readContentFromBody(openWireMessageBuffer, coreType, isCompressed);
+      if (bytes != null) {
+         ByteSequence content = new ByteSequence(bytes);
+         ActiveMQBuffer coreMessageBuffer = coreMessage.getBodyBuffer();
+         writeContentIntoBody(coreMessageBuffer, content, coreType, isCompressed);
+      }
+      coreMessage.putObjectProperty(FilterConstants.NATIVE_MESSAGE_ID,
+            openwireMessage.getObjectProperty(FilterConstants.NATIVE_MESSAGE_ID));
+      coreMessage.setMessageID(openwireMessage.getMessageID());
+      final String corrId = openwireMessage.getStringProperty(OpenWireMessageConverter.JMS_CORRELATION_ID_PROPERTY);
+      if (corrId != null) {
+         coreMessage.putStringProperty(OpenWireMessageConverter.JMS_CORRELATION_ID_PROPERTY, corrId);
+      }
+      final String groupId = openwireMessage.getStringProperty(org.apache.activemq.artemis.api.core.Message.HDR_GROUP_ID);
+      if (groupId != null) {
+         coreMessage.putStringProperty(org.apache.activemq.artemis.api.core.Message.HDR_GROUP_ID,
+               new SimpleString(groupId));
```

Comment:
`new SimpleString(groupId)` could be replaced by `coreMessageObjectPools.getGroupIdStringSimpleStringPool().getOrCreate(groupId)`

Reply:
+1 good suggestion, i think maybe we can do this optimisation later though, this is a large change, and improves things anyhow. e.g. why dont we get this PR in, and then break loose on perf later?